yarn-project/noir-contracts/src/contracts/test_contract/src/test_contract_interface.nr

```diff
@@ -0,0 +1,145 @@
+/* Autogenerated file, do not edit! */
+  
+use dep::std;
+use dep::aztec::context::Context;
+use dep::aztec::constants_gen::RETURN_VALUES_LENGTH;
+
+struct AStructTestCodeGenStruct {
+  amount: Field,
+  secretHash: Field,
+}
+
+struct ADeepStructTestCodeGenStruct {
+  aField: Field,
+  aBool: bool,
+  aNote: ANoteADeepStructTestCodeGenStruct,
+  manyNotes: [ManyNotesADeepStructTestCodeGenStruct;3],
+}
+
+struct ANoteADeepStructTestCodeGenStruct {
+  amount: Field,
+  secretHash: Field,
+}
+
+struct ManyNotesADeepStructTestCodeGenStruct {
+  amount: Field,
+  secretHash: Field,
+}
+
+struct TestContractInterface {
+  address: Field,
+}
+
+impl TestContractInterface {
+  fn at(address: Field) -> Self {
+      Self {
+          address,
+      }
+  }
+  
+  fn createL2ToL1MessagePublic(
+    self,
+    context: &mut Context,
+    amount: Field,
+    secretHash: Field
+  ) {
+    let mut serialised_args = [0; 2];
+    serialised_args[0] = amount;
+    serialised_args[1] = secretHash;
+
+    context.call_public_function(self.address, 0x1c031d17, serialised_args)
+  }
+  
+
+  fn createNullifierPublic(
+    self,
+    context: &mut Context,
+    amount: Field,
+    secretHash: Field
+  ) {
+    let mut serialised_args = [0; 2];
+    serialised_args[0] = amount;
+    serialised_args[1] = secretHash;
+
+    context.call_public_function(self.address, 0x0217ef40, serialised_args)
+  }
+  
+
+  fn getPortalContractAddress(
+    self,
+    context: &mut Context,
+    aztec_address: Field
+  ) -> [Field; RETURN_VALUES_LENGTH] {
+    let mut serialised_args = [0; 1];
+    serialised_args[0] = aztec_address;
+
+    context.call_private_function(self.address, 0xe5df1726, serialised_args)
+  }
+  
+
+  fn getPublicKey(
+    self,
+    context: &mut Context,
+    address: Field
+  ) -> [Field; RETURN_VALUES_LENGTH] {
+    let mut serialised_args = [0; 1];
+    serialised_args[0] = address;
+
+    context.call_private_function(self.address, 0x553aaad4, serialised_args)
+  }
+  
+
+  fn getThisAddress(
+    self,
+    context: &mut Context
+  ) -> [Field; RETURN_VALUES_LENGTH] {
+    let mut serialised_args = [0; 0];
+
+    context.call_private_function(self.address, 0xd3953822, serialised_args)
```

Comment:
Not needed for now, but if we continue with this kind of approach for generating contract interfaces, a developer shouldn't blindly trust these opaque hex function selectors. We probably need a way for a user who _uses_ one of these contract interfaces to _validate_ the correctness of the hex function selector.

Reply:
Fair point. Maybe we could include the selector in the generated ABIs? Or add a CLI command that outputs the list of functions and their selectors given a contract?

Reply:
I've opened an issue for the discussion to continue: https://github.com/AztecProtocol/aztec-packages/issues/1551